### ISSUE TITLE
Unused lifetime type parameters

### ISSUE THREAD

**Issue by regexident** [rust-lang/rust-clippy#877]
Leaving an unused type parameter  …

``` rust
struct Foo;

impl<T> Foo {
    fn foo() {}
}
```

… triggers the following error:

```
error: the type parameter `T` is not constrained
by the impl trait, self type, or predicates [E0207]
```

Leaving an unused lifetime parameter however …

``` rust
struct Bar;

impl<'a> Bar {
    fn bar() {}
}
```

does not.

We should lint these, shouldn't we? Or should I file a rustc bug report?

**Comment by llogiq** [comment]
There's the `unused_lifetimes` lint. Doesn't that catch your case?

**Comment by regexident** [comment]
No, apparently it doesn't.

This …
```rust
impl<'a> Bar {
    fn foo<'b>(_: u8) { }
    
    fn bar() {}
}
```

… only triggers one clippy warning. For `fn foo`:

```
src/lib.rs:4:12: 4:14 warning: this lifetime isn't used in the function definition,
#[warn(unused_lifetimes)] on by default
src/lib.rs:4     fn foo<'b>(_: u8) { }
                        ^~
src/lib.rs:4:12: 4:14 help: for further information visit
https://github.com/Manishearth/rust-clippy/wiki#unused_lifetimes
```

**Comment by llogiq** [comment]
Ok, so this should be an extension to the `unused_lifetimes` lint that should also check `ImplItems` (and probably `TraitImplItems`), which it currently apparently does not.

**Comment by birkenfeld** [comment]
I'll note that this behavior is currently pretty nice for writing macros that handle reference and non-reference types, because it saves you some contortions that come from the fact that there is no macro matcher for `Generics`.

Of course, for clippy, that doesn't really matter since we'd exclude macro expansions anyway.

**Comment by varkor** [comment]
The `unused_lifetimes` lint now captures the example in https://github.com/rust-lang/rust-clippy/issues/877#issuecomment-214039341, so it might now be sufficient.

**Comment by flip1995** [comment]
Is this a new addition to the rustc lint, that hasn't made it into the playground yet? Since it doesn't get linted on the [Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d672084029b7d21a3e7bf4333ace5308).

**Comment by varkor** [comment]
@flip1995: you were missing the `!`s — [here's a working link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=16ca93f2e8abef5d90afb7612585bebc).

**Comment by flip1995** [comment]
Ohhhhh, sorry :man_facepalming: Then this can be closed
